plenario/api/validator.py

```diff
@@ -0,0 +1,173 @@
+# ValidatorResult
+# ===============
+# Since much of the old code, formatting results into json, expected data as
+# attributes to the old ParamValidator object, it was convenient to use a named
+# tuple here that made a validator result compatible with response code.
+
+ValidatorResult = namedtuple('ValidatorResult', 'data errors warnings')
+
+
+# converters
+# ==========
+# Callables which are used to convert request arguments to their correct types.
+
+converters = {
+    'buffer': int,
+    'dataset': lambda x: MetaTable.get_by_dataset_name(x),
+    'dataset_name__in': lambda x: x.split(','),
+    'date__time_of_day_ge': int,
+    'date__time_of_day_le': int,
+    'obs_date__ge': lambda x: parser.parse(x).date(),
+    'obs_date__le': lambda x: parser.parse(x).date(),
+    'offset': int,
+    'resolution': int,
+    'geom': lambda x: make_fragment_str(extract_first_geometry_fragment(x)),
+    'shape': lambda x: get_shape_table(x)
+}
+
+
+def get_shape_table(shtable_name):
+    """A simple helper method to get a table that corresponds to a ShapeMetadata
+    record. This exists because while MetaTable.point_table exists, ShapeMetadata
+    doesn't have an analogous function.
+
+    :param shtable_name: name of shape table
+
+    :returns: table instance corresponding to ShapeMetadata record"""
+
+    return session.query(ShapeMetadata)\
+        .filter(ShapeMetadata.dataset_name == shtable_name)\
+        .first()\
+        .shape_table
+
+
+def convert(request_args):
+    """Convert a dictionary of arguments from strings to their types. How the
+    values are converted are specified by the converters dictionary defined
+    above.
+
+    :param request_args: dictionary of request arguments
+
+    :returns: converted dictionary"""
+
+    for key, value in request_args.items():
+        try:
+            request_args[key] = converters[key](value)
+        except:
+            # Failed transactions, which we do expect, can cause
+            # a DatabaseError with Postgres. Failing to rollback
+            # prevents further queries from being carried out.
+            session.rollback()
+
+
+def validate(validator_cls, request_args, defaults=True):
+    """Validate a dictionary of arguments. Substitute all missing fields with
+    defaults if not explicitly told to do otherwise.
+
+    :param validator_cls: what kind of validator to use
+    :param request_args: dictionary of arguments from a request object
+    :param defaults: if false, only substitute default values for provided args
+
+    :returns: ValidatorResult namedtuple"""
+
+    if defaults:
+        # this validator will return results/defaults for all fields
+        validator = validator_cls()
+    else:
+        # this validator will return results/defaults for only the fields
+        # provided in the request
+        validator = validator_cls(only=request_args.keys())
+
+    # convert string values to corresponding types so the validator can work
+    convert(request_args)
```

Comment:
I find it confusing that we pass in `request_args` and mutate it, but the caller never uses `request_args` after the fact.